src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1

```diff
@@ -0,0 +1,205 @@
+              {
+                 Set-AzKeyVaultAccessPolicy -VaultName $KeyvaultName -ObjectId $vault.IdentityPrincipalId -PermissionsToSecrets Get,List
+              }
+
+              else
+              {
+                  Write-Host "Assigning RBAC Policies"
+
+                  $CheckPermission = $AllRoles
+                  | Where-Object { ($_.Scope -eq $KeyvaultId -or $_.Scope -eq $KeyvaultRG -or  $_.Scope -eq $KeyvaultSubscription) -and $_.RoleDefinitionName -eq $Permission}
+              
+                  if($CheckPermission -ne $null)
+                  {
+                      Write-Host "Required RBAC Permissions Already Assigned."
+                  }
+
+                  else
+                  {
+                      $MissingRolesInitially = $true
+
+                      Write-Host "Assigning Required RBAC Permissions"
+
+                      if($PermissionsScope -eq "Resource")
+                      {
+                          AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $KeyvaultId
+                      }
+              
+                      elseif($PermissionsScope -eq "ResourceGroup")
+                      {
+                          AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $KeyvaultRG
+                      }
+
+                      elseif($PermissionsScope -eq "Subscription")
+                      { 
+                          AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $KeyvaultSubscriptionName
+                      }                 
+
+                      Write-Host "Assigned Required RBAC Permissions"
+                  }
+              }
+          }
+               
+          foreach($Permission in $manifest.datasourcePermissions)
+          {
+              $CheckPermission = $AllRoles
+              | Where-Object { ($_.Scope -eq $DataSourceId -or $_.Scope -eq $ResourceRG -or  $_.Scope -eq $SubscriptionName) -and $_.RoleDefinitionName -eq $Permission}
+              
+              if($CheckPermission -ne $null)
+              {
+                  Write-Host "Required Resource Permissions Already Assigned."
+              }
+
+              else
+              {
+                  $MissingRolesInitially = $true
+
+                  Write-Host "Assigning Required Resource Permissions"
+
+                  if($PermissionsScope -eq "Resource")
+                  {
+                     AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $DatasourceId
+                  }
+              
+                  elseif($PermissionsScope -eq "ResourceGroup")
+                  {
+                     AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $ResourceRG
+                  }
+
+                  elseif($PermissionsScope -eq "Subscription")
```

Comment:
This code can be shortened with use of a helper function that on providing the armid and the scope returns the corresponding scope id.